### PR TITLE
[Spring] Add new params for custom actuator 

### PR DESCRIPTION
**Pull request by Netyyyy**
---

 Add new params for custom actuator 

### Related command

Add argument `--custom-actuator-port`, `--custom-actuator-path` in `spring app update` and `spring app deploy` and `spring app deployment create`

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)
- [ ] My extension version conforms to the [Extension version schema](https://github.com/Azure/azure-cli/blob/release/doc/extensions/versioning_guidelines.md)

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
Spring

File: src/spring/azext_spring/_deployment_factory.py [hunk 1]
```diff
@@ -104,13 +107,22 @@ def _get_env(self, env=None, **_):
...
+        if any([config_file_patterns, custom_actuator_port, custom_actuator_path]):
+            addon_configs = {}
+            if config_file_patterns is not None:
+                addon_configs = {
+                    APPLICATION_CONFIGURATION_SERVICE_NAME: {
+                        APPLICATION_CONFIGURATION_SERVICE_PROPERTY_PATTERN: config_file_patterns
+                    }
                 }
-            }
+            app_live_view = {}
```
Comment (saragluna) [inline src/spring/azext_spring/_deployment_factory.py:119]: will this work?

```python
    if any([custom_actuator_port, custom_actuator_path]): 
          addon_configs[APP_LIVE_VIEW] = {
              ACTUATOR_PORT: custom_actuator_port,
              ACTUATOR_PATH: custom_actuator_path
          }
```

File: src/spring/azext_spring/_deployment_factory.py [hunk 2]
```diff
@@ -238,20 +250,38 @@ def deployment_settings_options_from_resource(original):
...
+        'custom_actuator_path': _get_origin_custom_actuator_path(original.properties.deployment_settings.addon_configs)
     }
     if original.properties.deployment_settings.container_probe_settings is not None:
         options['disable_probe'] = original.properties.deployment_settings.container_probe_settings.disable_probe
     return options
 
 
 def _get_origin_config_file_patterns(origin_addon_configs):
-    if origin_addon_configs:
+    if APPLICATION_CONFIGURATION_SERVICE_NAME in origin_addon_configs:
```
Comment (saragluna) [inline src/spring/azext_spring/_deployment_factory.py:263]: Will this work if the origin_addon_configs is None?

Comment (Netyyyy) [reply src/spring/azext_spring/_deployment_factory.py:263]: not work, it only checks the values from origin app deployment

File: src/spring/azext_spring/_deployment_factory.py [hunk 3]
```diff
@@ -238,20 +248,38 @@ def deployment_settings_options_from_resource(original):
...
+
+
+def _get_origin_custom_actuator_path(origin_addon_configs):
+    if APP_LIVE_VIEW in origin_addon_configs:
+        alv_addon = origin_addon_configs.get(APP_LIVE_VIEW)
+        if alv_addon.get(ACTUATOR_PATH):
+            return alv_addon.get(ACTUATOR_PATH)
+    return None
+
+
```
Comment (yuwzho) [inline src/spring/azext_spring/_deployment_factory.py:282]: Can you consider write a general method in util.py that can look up a value from a given dict then all add_on file can leverage it?  Then you can call it like `_get_value_from_dict(origin_addon_config, APP_LIVE_VIEW, ACTUATOR_PATH)`
```python
def _get_value_from_dict(input_dict, *keys):
    for key in keys:
        if not isinstance(input_dict, dict):
            return None
        input_dict = input_dict.get(key)
    return input_dict
```

File: src/spring/azext_spring/_validators_enterprise.py [hunk 4]
```diff
@@ -714,3 +714,19 @@ def _get_eactly_one_application_configuration_service_resource_id(cmd, resource_
     if len(acs_resources) > 1:
         raise ClientRequestError('App cannot bind to multiple application configuration services.')
     return acs_resources[0].id
+
+
+def validate_custom_actuator_port(cmd, namespace):
+    only_support_enterprise(cmd, namespace)
+    if namespace.custom_actuator_port:
+        if not isinstance(namespace.custom_actuator_port, int):
+            raise InvalidArgumentValueError("--custom-actuator-port should be a number.")
```
Comment (yuwzho) [inline src/spring/azext_spring/_validators_enterprise.py:723]:
```suggestion
```
The type already check it is int.

File: src/spring/azext_spring/_validators_enterprise.py [hunk 5]
```diff
@@ -714,3 +714,19 @@ def _get_eactly_one_application_configuration_service_resource_id(cmd, resource_
...
+        if namespace.custom_actuator_port < 0:
```
Comment (yuwzho) [inline src/spring/azext_spring/_validators_enterprise.py:724]:
```suggestion
        if namespace.custom_actuator_port <= 0:
```

File: src/spring/azext_spring/_validators_enterprise.py [hunk 6]
```diff
@@ -714,3 +714,19 @@ def _get_eactly_one_application_configuration_service_resource_id(cmd, resource_
...
+            raise ArgumentUsageError("--custom-actuator-port must be greater than or equal to 0")
+
+
+def validate_custom_actuator_path(cmd, namespace):
+    only_support_enterprise(cmd, namespace)
+    if namespace.custom_actuator_path:
+        if not isinstance(namespace.custom_actuator_path, str):
+            raise InvalidArgumentValueError("--custom-actuator-path should be a string")
```
Comment (yuwzho) [inline src/spring/azext_spring/_validators_enterprise.py:732]:
```suggestion
```

The type already checked it is a string

File: src/spring/azext_spring/tests/latest/test_asa_app.py [hunk 7]
```diff
@@ -322,6 +322,32 @@ def test_app_deploy_enterprise(self, file_mock):
         self.assertEqual({'applicationConfigurationService': {'configFilePatterns': 'my-pattern'}},
                          resource.properties.deployment_settings.addon_configs)
 
+    @mock.patch('azext_spring._deployment_uploadable_factory.FileUpload.upload_and_build')
```
Comment (yuwzho) [inline src/spring/azext_spring/tests/latest/test_asa_app.py:325]: Please also add a  test that config `config_file_patterns` and check the `addon_configs`.

File: src/spring/azext_spring/_validators_enterprise.py [hunk 8]
```diff
@@ -714,3 +714,10 @@ def _get_eactly_one_application_configuration_service_resource_id(cmd, resource_
     if len(acs_resources) > 1:
         raise ClientRequestError('App cannot bind to multiple application configuration services.')
     return acs_resources[0].id
+
+
+def validate_custom_actuator_port(cmd, namespace):
+    only_support_enterprise(cmd, namespace)
+    if namespace.custom_actuator_port:
+        if namespace.custom_actuator_port <= 0:
+            raise ArgumentUsageError("--custom-actuator-port must be greater than or equal to 0")
```
Comment (yuwzho) [inline src/spring/azext_spring/_validators_enterprise.py:723]:
```suggestion
            raise ArgumentUsageError("--custom-actuator-port must be greater than 0")
```

**Review by yuwzho — APPROVED** [review]

**Comment by yonzhan** [comment]
Please fix CI issues.

**Comment by Netyyyy** [comment]
@zhoxing-ms @yonzhan , please help review, thanks. And The failed CI is warning about the original code. Let us keep this PR clean for the new feature and we commit the incremental code passed all CI check.

**Comment by Netyyyy** [comment]
@yonzhan @zhoxing-ms @yanzhudd please help approve, thanks

File: src/spring/HISTORY.md [hunk 9]
```diff
@@ -1,5 +1,9 @@
 Release History
 ===============
+1.22.0
+---
+* Add arguments `--custom-actuator-port` and `--custom-actuator-path` in `spring app update`, `spring app deploy` and `spring app deployment create` to support custom actuator.
```
Comment (yanzhudd) [inline src/spring/HISTORY.md:5]: It is better to split this changelog into multiple logs.

File: src/spring/azext_spring/_params.py [hunk 10]
```diff
@@ -547,6 +548,10 @@ def prepare_logs_argument(c):
                        help="(Enterprise Tier Only) Config file patterns separated with \',\' to decide which patterns "
                             "of Application Configuration Service will be used. Use '\"\"' to clear existing configurations.",
                        validator=validate_config_file_patterns)
+            c.argument('custom_actuator_port', type=int,
+                       help='(Enterprise Tier Only) Custom actuator port for the app. Default to 8080.', validator=validate_custom_actuator_port)
+            c.argument('custom_actuator_path', type=str,
+                       help='(Enterprise Tier Only) Custom actuator path for the app. Default to "/actuator".', validator=validate_custom_actuator_path)
```
Comment (yanzhudd) [inline src/spring/azext_spring/_params.py:554]: It is not necessary to specify `type=str`, as the type is set to str be default.
```suggestion
            c.argument('custom_actuator_path', 
                       help='(Enterprise Tier Only) Custom actuator path for the app. Default to "/actuator".', validator=validate_custom_actuator_path)
```

**Comment by yanzhudd** [comment]
Do you want to release a new version for this change? If so, please update the version in setup.py as well.

**Review by yanzhudd — APPROVED** [review]

**Pull request merged**